Unable to reuse buckets names from buckets previously deleted

### Environment info
- NooBaa Version: 5.15.4
- Platform: standalone

### Actual behavior
1. Create a bucket bucket1 with filesystempath /mnt/gpfs1/tes/test1
2. Delete the filesystempath /mnt/gpfs1/tes/test1
3. Now try to delete the bucket bucket1. It fails but bucket list displays it.

# noobaa-cli bucket delete bucket1 2>/dev/null
{
  "error": {
    "code": "NoSuchBucket",
    "message": "Bucket does not exist"
  }
}
# noobaa-cli bucket list 2>/dev/null
{
  "response": {
    "code": "BucketList",
    "reply": [
      {
        "name": "bucket1"
      }
    ]
  }
}



### Expected behavior
1. It should error out properly instead of saying bucket doesnot exist it should say bucket path missing

### Steps to reproduce
1.

### More information - Screenshots / Logs / Other output

@romayalon I see an issue on S3 flow, and also we would need to fix the error code.
Do you prefer to open an issue for this or continue on this issue?

Partial output:
```
2024-07-16 14:57:29.395996 [PID-79847/TID-259] [L1] FS::FSWorker::OnError: Stat _path=/tmp/nsfs_root1/bucket-1001-16-07  error.Message()=No such file or directory
Jul-16 14:57:29.398 [nsfs/79847] [ERROR] core.sdk.bucketspace_fs:: BucketSpaceFS: delete_bucket: error [Error: No such file or directory] { code: 'ENOENT', rpc_code: 'NO_SUCH_OBJECT' }
Jul-16 14:57:29.398 [nsfs/79847] [ERROR] core.endpoint.s3.s3_rest:: S3 ERROR <?xml version="1.0" encoding="UTF-8"?><Error><Code>NoSuchKey</Code><Message>The specified key does not exist.</Message><Resource>/bucket-1001-16-07</Resource><RequestId>lyoczl4b-db24b2-oza</RequestId></Error> DELETE /bucket-1001-16-07 {"host":"localhost:6443","accept-encoding":"identity","user-agent":"aws-cli/2.17.11 md/awscrt#0.20.11 ua/2.0 os/macos#23.4.0 md/arch#arm64 lang/python#3.11.9 md/pyimpl#CPython cfg/retry-mode#standard md/installer#source md/prompt#off md/command#s3.rb","x-amz-date":"20240716T115729Z","x-amz-content-sha256":"e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855","authorization":"AWS4-HMAC-SHA256 Credential=<>/20240716/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=163e27c0cf9f108363fef1f54034a505f6f4f43bae6d52d604fa16026ee356a2","content-length":"0"} Error: No such file or directory
```

In the flow, it fails on: `delete_bucket` (bucketspace_fs) -> ns.delete_uls (namespace_fs) -> fails when it tries to list -> error translated from `ENOENT` to `err.rpc_code = 'NO_SUCH_OBJECT'` in namespace_fs and then we don't translate it again in bucketspace_fs again.

Steps to reproduce:
1. Create the root user account with the CLI: `sudo node src/cmd/manage_nsfs account add --name shira-1001 --new_buckets_path /tmp/nsfs_root1 --access_key <access-key> --secret_key <secret-key> --uid <uid> --gid <gid>`
Note: before creating the account need to give permission to the `new_buckets_path`: `chmod 777 /tmp/nsfs_root1`.
2. Start the NSFS server with: `sudo node src/cmd/nsfs --debug 5`
Note: before starting the server please add this line: `process.env.NOOBAA_LOG_LEVEL = 'nsfs';` in the endpoint.js (before the condition `if (process.env.NOOBAA_LOG_LEVEL) {`)
3. Create the alias for S3 service: `alias s3-nc-user-1='AWS_ACCESS_KEY_ID=<access-key> AWS_SECRET_ACCESS_KEY=<secret-key> aws --no-verify-ssl --endpoint-url https://localhost:6443'`.
4. Create the bucket using S3: `s3-nc-user-1-s3 s3 mb s3://bucket-1001-16-07`
5. Remove the directory that was created for the bucket: `sudo rmdir /tmp/nsfs_root1/bucket-1001-16-07`
6. Try to delete the bucket: `s3-nc-user-1-s3 s3 rb s3://bucket-1001-16-07`, the output: `remove_bucket failed: s3://bucket-1001-16-07 An error occurred (NoSuchKey) when calling the DeleteBucket operation: The specified key does not exist.`

--- a/src/sdk/accountspace_fs.js
+++ b/src/sdk/accountspace_fs.js
@@ -21,11 +21,6 @@ const { get_symlink_config_file_path, get_config_file_path, get_config_data,
 const nc_mkm = require('../manage_nsfs/nc_master_key_manager').get_instance();
 const { account_cache } = require('./object_sdk');
 
-const entity_enum = Object.freeze({
-    USER: 'USER',
-    ACCESS_KEY: 'ACCESS_KEY',
-});
-
 // TODO - rename (the typo), move and reuse in manage_nsfs
 const acounts_dir_relative_path = '../accounts/';
 
@@ -87,7 +82,7 @@ class AccountSpaceFS {
             };
         } catch (err) {
             dbg.error(`AccountSpaceFS.${action} error`, err);
-            throw this._translate_error_codes(err, entity_enum.USER);
+            throw native_fs_utils.translate_error_codes(err, native_fs_utils.entity_enum.USER);
         }
     }
 
@@ -125,7 +120,7 @@ class AccountSpaceFS {
             };
         } catch (err) {
             dbg.error(`AccountSpaceFS.${action} error`, err);
-            throw this._translate_error_codes(err, entity_enum.USER);
+            throw native_fs_utils.translate_error_codes(err, native_fs_utils.entity_enum.USER);
         }
     }
 
@@ -179,7 +174,7 @@ class AccountSpaceFS {
             };
         } catch (err) {
             dbg.error(`AccountSpaceFS.${action} error`, err);
-            throw this._translate_error_codes(err, entity_enum.USER);
+            throw native_fs_utils.translate_error_codes(err, native_fs_utils.entity_enum.USER);
         }
     }
 
@@ -213,7 +208,7 @@ class AccountSpaceFS {
             await native_fs_utils.delete_config_file(this.fs_context, this.accounts_dir, account_config_path);
         } catch (err) {
             dbg.error(`AccountSpaceFS.${action} error`, err);
-            throw this._translate_error_codes(err, entity_enum.USER);
+            throw native_fs_utils.translate_error_codes(err, native_fs_utils.entity_enum.USER);
         }
     }
 
@@ -237,7 +232,7 @@ class AccountSpaceFS {
         return { members, is_truncated };
         } catch (err) {
             dbg.error(`AccountSpaceFS.${action} error`, err);
-            throw this._translate_error_codes(err, entity_enum.USER);
+            throw native_fs_utils.translate_error_codes(err, native_fs_utils.entity_enum.USER);
         }
     }
 
@@ -297,7 +292,7 @@ class AccountSpaceFS {
             };
         } catch (err) {
             dbg.error(`AccountSpaceFS.${action} error`, err);
-            throw this._translate_error_codes(err, entity_enum.ACCESS_KEY);
+            throw native_fs_utils.translate_error_codes(err, native_fs_utils.entity_enum.ACCESS_KEY);
         }
     }
 
@@ -329,7 +324,7 @@ class AccountSpaceFS {
             };
         } catch (err) {
             dbg.error('AccountSpaceFS.get_access_key_last_used error', err);
-            throw this._translate_error_codes(err, entity_enum.ACCESS_KEY);
+            throw native_fs_utils.translate_error_codes(err, native_fs_utils.entity_enum.ACCESS_KEY);
         }
     }
 
@@ -380,7 +375,7 @@ class AccountSpaceFS {
             this._clean_account_cache(requested_account);
         } catch (err) {
             dbg.error(`AccountSpaceFS.${action} error`, err);
-            throw this._translate_error_codes(err, entity_enum.ACCESS_KEY);
+            throw native_fs_utils.translate_error_codes(err, native_fs_utils.entity_enum.ACCESS_KEY);
         }
     }
 
@@ -426,7 +421,7 @@ class AccountSpaceFS {
             this._clean_account_cache(requested_account);
         } catch (err) {
             dbg.error(`AccountSpaceFS.${action} error`, err);
-            throw this._translate_error_codes(err, entity_enum.ACCESS_KEY);
+            throw native_fs_utils.translate_error_codes(err, native_fs_utils.entity_enum.ACCESS_KEY);
         }
     }
 
@@ -460,25 +455,13 @@ class AccountSpaceFS {
             return { members, is_truncated, username: name_for_access_key };
         } catch (err) {
             dbg.error(`AccountSpaceFS.${action} error`, err);
-            throw this._translate_error_codes(err, entity_enum.ACCESS_KEY);
+            throw native_fs_utils.translate_error_codes(err, native_fs_utils.entity_enum.ACCESS_KEY);
         }
     }
 
     ////////////////////////
     // INTERNAL FUNCTIONS //
     ////////////////////////
-
-    // this function was copied from namespace_fs and bucketspace_fs
-    // It is a fallback that we use, but might be not accurate
-    _translate_error_codes(err, entity) {
-        if (err.rpc_code) return err;
-        if (err.code === 'ENOENT') err.rpc_code = `NO_SUCH_${entity}`;
-        if (err.code === 'EEXIST') err.rpc_code = `${entity}_ALREADY_EXISTS`;
-        if (err.code === 'EPERM' || err.code === 'EACCES') err.rpc_code = 'UNAUTHORIZED';
-        if (err.code === 'IO_STREAM_ITEM_TIMEOUT') err.rpc_code = 'IO_STREAM_ITEM_TIMEOUT';
-        if (err.code === 'INTERNAL_ERROR') err.rpc_code = 'INTERNAL_ERROR';
-        return err;
-    }
 
      _get_account_config_path(name) {
         return get_config_file_path(this.accounts_dir, name);
@@ -572,7 +555,7 @@ class AccountSpaceFS {
         const basic_message = `User: ${arn_for_requesting_account} is not authorized to perform:` +
         `${full_action_name} on resource: `;
         let message_with_details;
-        if (entity === entity_enum.USER) {
+        if (entity === native_fs_utils.entity_enum.USER) {
             let user_message;
             if (action === 'list_access_keys') {
                 user_message = `user ${requesting_account.name.unwrap()}`;
@@ -581,7 +564,7 @@ class AccountSpaceFS {
             }
             message_with_details = basic_message +
             `${user_message} because no identity-based policy allows the ${full_action_name} action`;
-        } else { // entity_enum.ACCESS_KEY
+        } else { // native_fs_utils.entity_enum.ACCESS_KEY
             message_with_details = basic_message + `access key ${details.access_key}`;
         }
         const { code, http_code, type } = IamError.AccessDeniedException;
@@ -658,7 +641,7 @@ class AccountSpaceFS {
         if (!is_root_account) {
             dbg.error(`AccountSpaceFS.${action} requesting account is not a root account`,
                 requesting_account);
-            this._throw_access_denied_error(action, requesting_account, user_details, entity_enum.USER);
+            this._throw_access_denied_error(action, requesting_account, user_details, native_fs_utils.entity_enum.USER);
         }
     }
 
@@ -895,7 +878,7 @@ class AccountSpaceFS {
             dbg.error(`AccountSpaceFS.${action} requesting account is neither a root account ` +
             `nor user requester on himself`,
             requesting_account);
-            this._throw_access_denied_error(action, requesting_account, { username }, entity_enum.USER);
+            this._throw_access_denied_error(action, requesting_account, { username }, native_fs_utils.entity_enum.USER);
         }
         return requester;
     }

--- a/src/sdk/bucketspace_fs.js
+++ b/src/sdk/bucketspace_fs.js
@@ -21,7 +21,8 @@ const { CONFIG_SUBDIRS } = require('../manage_nsfs/manage_nsfs_constants');
 
 const KeysSemaphore = require('../util/keys_semaphore');
 const { get_umasked_mode, isDirectory, validate_bucket_creation,
-    create_config_file, delete_config_file, get_bucket_tmpdir_full_path, folder_delete } = require('../util/native_fs_utils');
+    create_config_file, delete_config_file, get_bucket_tmpdir_full_path, folder_delete,
+    entity_enum, translate_error_codes } = require('../util/native_fs_utils');
 const NoobaaEvent = require('../manage_nsfs/manage_nsfs_events_utils').NoobaaEvent;
 const { anonymous_access_key } = require('./object_sdk');
 
@@ -84,16 +85,6 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
         } catch (err) {
             return false;
         }
-    }
-
-    _translate_bucket_error_codes(err) {
-        if (err.rpc_code) return err;
-        if (err.code === 'ENOENT') err.rpc_code = 'NO_SUCH_BUCKET';
-        if (err.code === 'EEXIST') err.rpc_code = 'BUCKET_ALREADY_EXISTS';
-        if (err.code === 'EPERM' || err.code === 'EACCES') err.rpc_code = 'UNAUTHORIZED';
-        if (err.code === 'IO_STREAM_ITEM_TIMEOUT') err.rpc_code = 'IO_STREAM_ITEM_TIMEOUT';
-        if (err.code === 'INTERNAL_ERROR') err.rpc_code = 'INTERNAL_ERROR';
-        return err;
     }
 
     async read_account_by_access_key({ access_key }) {
@@ -178,7 +169,7 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
             }
             return bucket;
         } catch (err) {
-            const rpc_error = this._translate_bucket_error_codes(err);
+            const rpc_error = translate_error_codes(err, entity_enum.BUCKET);
             if (err.rpc_code === 'INVALID_SCHEMA') err.rpc_code = 'INVALID_BUCKET_STATE';
             new NoobaaEvent(NoobaaEvent[rpc_error.rpc_code]).create_event(name, {bucket_name: name}, err);
             throw rpc_error;
@@ -222,7 +213,7 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
                 dbg.error('BucketSpaceFS: root dir not found', err, this.bucket_schema_dir);
                 throw new S3Error(S3Error.NoSuchBucket);
             }
-            throw this._translate_bucket_error_codes(err);
+            throw translate_error_codes(err, entity_enum.BUCKET);
         }
 
         const account = object_sdk.requesting_account;
@@ -302,7 +293,7 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
                 await create_config_file(this.fs_context, this.bucket_schema_dir, bucket_config_path, bucket_config);
             } catch (err) {
                 new NoobaaEvent(NoobaaEvent.BUCKET_CREATION_FAILED).create_event(name, {bucket_name: name}, err);
-                throw this._translate_bucket_error_codes(err);
+                throw translate_error_codes(err, entity_enum.BUCKET);
             }
 
             // create bucket's underlying storage directory
@@ -313,7 +304,7 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
                 dbg.error('BucketSpaceFS: create_bucket could not create underlying directory - nsfs, deleting bucket', err);
                 new NoobaaEvent(NoobaaEvent.BUCKET_DIR_CREATION_FAILED).create_event(name, {bucket: name, path: bucket_storage_path}, err);
                 await nb_native().fs.unlink(this.fs_context, bucket_config_path);
-                throw this._translate_bucket_error_codes(err);
+                throw translate_error_codes(err, entity_enum.BUCKET);
             }
         });
     }
@@ -359,13 +350,27 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
                 if (!namespace_bucket_config) throw new RpcError('INTERNAL_ERROR', 'Invalid Bucket configuration');
 
                 if (namespace_bucket_config.should_create_underlying_storage) {
-                    // 1. delete underlying storage
+                    // 1. delete underlying storage (ULS = Underline Storage)
                     dbg.log1('BucketSpaceFS.delete_bucket: deleting uls', this.fs_root, namespace_bucket_config.write_resource.path);
                     const bucket_storage_path = path.join(this.fs_root, namespace_bucket_config.write_resource.path); // includes write_resource.path + bucket name (s3 flow)
-                    await ns.delete_uls({ name, full_path: bucket_storage_path }, object_sdk);
+                    try {
+                        await ns.delete_uls({ name, full_path: bucket_storage_path }, object_sdk);
+                    } catch (err) {
+                        dbg.warn('delete_bucket: bucket name', name, 'with bucket_storage_path', bucket_storage_path,
+                            'got an error while trying to delete_uls', err);
+                        // in case the ULS was deleted - we will continue
+                        if (err.rpc_code !== 'NO_SUCH_BUCKET') throw err;
+                    }
                 } else {
                     // 2. delete only bucket tmpdir
-                    const list = await ns.list_objects({ ...params, bucket: name, limit: 1 }, object_sdk);
+                    let list;
+                    try {
+                        list = await ns.list_objects({ ...params, bucket: name, limit: 1 }, object_sdk);
+                    } catch (err) {
+                        dbg.warn('delete_bucket: bucket name', name, 'got an error while trying to list_objects', err);
+                        // in case the ULS was deleted - we will continue
+                        if (err.rpc_code !== 'NO_SUCH_BUCKET') throw err;
+                    }
                     if (list && list.objects && list.objects.length > 0) throw new RpcError('NOT_EMPTY', 'underlying directory has files in it');
                     const bucket_tmpdir_path = get_bucket_tmpdir_full_path(namespace_bucket_config.write_resource.path, bucket._id);
                     await folder_delete(bucket_tmpdir_path, this.fs_context, true);
@@ -375,10 +380,10 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
                 await delete_config_file(this.fs_context, this.bucket_schema_dir, bucket_config_path);
                 new NoobaaEvent(NoobaaEvent.BUCKET_DELETE).create_event(name, { bucket_name: name });
             } catch (err) {
-                dbg.error('BucketSpaceFS: delete_bucket: error', err);
+                dbg.error('BucketSpaceFS: delete_bucket: bucket name', name, 'error', err);
                 new NoobaaEvent(NoobaaEvent.BUCKET_DELETE_FAILED).create_event(name,
                     { bucket_name: name, bucket_path: bucket_config_path }, err);
-                throw this._translate_bucket_error_codes(err);
+                throw translate_error_codes(err, entity_enum.BUCKET);
             }
         });
     }
@@ -426,7 +431,7 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
                 }
             );
         } catch (err) {
-            throw this._translate_bucket_error_codes(err);
+            throw translate_error_codes(err, entity_enum.BUCKET);
         }
     }
 
@@ -484,7 +489,7 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
                 }
             );
         } catch (err) {
-            throw this._translate_bucket_error_codes(err);
+            throw translate_error_codes(err, entity_enum.BUCKET);
         }
     }
 
@@ -508,7 +513,7 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
                 }
             );
         } catch (err) {
-            throw this._translate_bucket_error_codes(err);
+            throw translate_error_codes(err, entity_enum.BUCKET);
         }
     }
 
@@ -521,7 +526,7 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
             const bucket = JSON.parse(data.toString());
             return bucket.logging;
         } catch (err) {
-            throw this._translate_bucket_error_codes(err);
+            throw translate_error_codes(err, entity_enum.BUCKET);
         }
     }
 
@@ -550,7 +555,7 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
                 { mode: get_umasked_mode(config.BASE_MODE_CONFIG_FILE) }
             );
         } catch (err) {
-            throw this._translate_bucket_error_codes(err);
+            throw translate_error_codes(err, entity_enum.BUCKET);
         }
     }
 
@@ -563,7 +568,7 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
             const bucket = JSON.parse(data.toString());
             return bucket.encryption;
         } catch (err) {
-            throw this._translate_bucket_error_codes(err);
+            throw translate_error_codes(err, entity_enum.BUCKET);
         }
     }
 
@@ -586,7 +591,7 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
                 { mode: get_umasked_mode(config.BASE_MODE_CONFIG_FILE) }
             );
         } catch (err) {
-            throw this._translate_bucket_error_codes(err);
+            throw translate_error_codes(err, entity_enum.BUCKET);
         }
     }
 
@@ -614,7 +619,7 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
                 { mode: get_umasked_mode(config.BASE_MODE_CONFIG_FILE) }
             );
         } catch (err) {
-            throw this._translate_bucket_error_codes(err);
+            throw translate_error_codes(err, entity_enum.BUCKET);
         }
     }
 
@@ -637,7 +642,7 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
                 { mode: get_umasked_mode(config.BASE_MODE_CONFIG_FILE) }
             );
         } catch (err) {
-            throw this._translate_bucket_error_codes(err);
+            throw translate_error_codes(err, entity_enum.BUCKET);
         }
     }
 
@@ -654,7 +659,7 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
             const bucket = JSON.parse(data.toString());
             return { website: bucket.website };
         } catch (err) {
-            throw this._translate_bucket_error_codes(err);
+            throw translate_error_codes(err, entity_enum.BUCKET);
         }
     }
 
@@ -685,7 +690,7 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
                 { mode: get_umasked_mode(config.BASE_MODE_CONFIG_FILE) }
             );
         } catch (err) {
-            throw this._translate_bucket_error_codes(err);
+            throw translate_error_codes(err, entity_enum.BUCKET);
         }
     }
 
@@ -708,7 +713,7 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
                 { mode: get_umasked_mode(config.BASE_MODE_CONFIG_FILE) }
             );
         } catch (err) {
-            throw this._translate_bucket_error_codes(err);
+            throw translate_error_codes(err, entity_enum.BUCKET);
         }
     }
 
@@ -722,7 +727,7 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
                 policy: bucket_policy_info.s3_policy
             };
         } catch (err) {
-            throw this._translate_bucket_error_codes(err);
+            throw translate_error_codes(err, entity_enum.BUCKET);
         }
     }
 

--- a/src/sdk/namespace_fs.js
+++ b/src/sdk/namespace_fs.js
@@ -891,7 +891,7 @@ class NamespaceFS {
             }
             return res;
         } catch (err) {
-            throw this._translate_object_error_codes(err);
+            throw native_fs_utils.translate_error_codes(err, native_fs_utils.entity_enum.OBJECT);
         }
     }
 
@@ -926,7 +926,7 @@ class NamespaceFS {
             if (this._should_update_issues_report(params, file_path, err)) {
                 this.run_update_issues_report(object_sdk, err);
             }
-            throw this._translate_object_error_codes(err);
+            throw native_fs_utils.translate_error_codes(err, native_fs_utils.entity_enum.OBJECT);
         }
     }
 
@@ -1085,7 +1085,7 @@ class NamespaceFS {
             //failed to get object
             new NoobaaEvent(NoobaaEvent.OBJECT_STREAM_GET_FAILED).create_event(params.key,
                                     {bucket_path: this.bucket_path, object_name: params.key}, err);
-            throw this._translate_object_error_codes(err);
+            throw native_fs_utils.translate_error_codes(err, native_fs_utils.entity_enum.OBJECT);
 
         } finally {
             try {
@@ -1151,7 +1151,7 @@ class NamespaceFS {
             new NoobaaEvent(NoobaaEvent.OBJECT_UPLOAD_FAILED).create_event(params.key,
                 {bucket_path: this.bucket_path, object_name: params.key}, err);
             dbg.warn('NamespaceFS: upload_object buffer pool cleanup error', err);
-            throw this._translate_object_error_codes(err);
+            throw native_fs_utils.translate_error_codes(err, native_fs_utils.entity_enum.OBJECT);
         } finally {
             try {
                 if (upload_params && upload_params.target_file) await upload_params.target_file.close(fs_context);
@@ -1581,7 +1581,7 @@ class NamespaceFS {
             );
             return { obj_id: params.obj_id };
         } catch (err) {
-            throw this._translate_object_error_codes(err);
+            throw native_fs_utils.translate_error_codes(err, native_fs_utils.entity_enum.OBJECT);
         }
     }
 
@@ -1653,7 +1653,7 @@ class NamespaceFS {
             return upload_info;
         } catch (err) {
             this.run_update_issues_report(object_sdk, err);
-            throw this._translate_object_error_codes(err);
+            throw native_fs_utils.translate_error_codes(err, native_fs_utils.entity_enum.OBJECT);
         } finally {
             await native_fs_utils.finally_close_files(fs_context, [target_file, part_md_file]);
         }
@@ -1693,7 +1693,7 @@ class NamespaceFS {
                 storage_class: create_multipart_upload_params.storage_class
             };
         } catch (err) {
-            throw this._translate_object_error_codes(err);
+            throw native_fs_utils.translate_error_codes(err, native_fs_utils.entity_enum.OBJECT);
         }
     }
 
@@ -1798,7 +1798,7 @@ class NamespaceFS {
             return upload_info;
         } catch (err) {
             dbg.error(err);
-            throw this._translate_object_error_codes(err);
+            throw native_fs_utils.translate_error_codes(err, native_fs_utils.entity_enum.OBJECT);
         } finally {
             await this.complete_object_upload_finally(undefined, [...part_size_to_fd_map.values()], target_file, fs_context);
         }
@@ -1849,7 +1849,7 @@ class NamespaceFS {
             }
             return res || {};
         } catch (err) {
-            throw this._translate_object_error_codes(err);
+            throw native_fs_utils.translate_error_codes(err, native_fs_utils.entity_enum.OBJECT);
         }
     }
 
@@ -1890,7 +1890,7 @@ class NamespaceFS {
             }
             return res;
         } catch (err) {
-            throw this._translate_object_error_codes(err);
+            throw native_fs_utils.translate_error_codes(err, native_fs_utils.entity_enum.OBJECT);
         }
     }
 
@@ -1920,7 +1920,7 @@ class NamespaceFS {
             await nb_native().fs.checkAccess(fs_context, this.bucket_path);
             this.versioning = versioning;
         } catch (err) {
-            throw this._translate_object_error_codes(err);
+            throw native_fs_utils.translate_error_codes(err, native_fs_utils.entity_enum.BUCKET);
         }
     }
 
@@ -1953,7 +1953,7 @@ class NamespaceFS {
             }
         } catch (err) {
             dbg.error(`NamespaceFS.get_object_tagging: failed in dir ${file_path} with error: `, err);
-            throw this._translate_object_error_codes(err);
+            throw native_fs_utils.translate_error_codes(err, native_fs_utils.entity_enum.OBJECT);
         }
         dbg.log0('NamespaceFS.get_object_tagging: return tagging ', tag_set, 'file_path :', file_path);
         return { tagging: tag_set };
@@ -1972,7 +1972,7 @@ class NamespaceFS {
             await this._clear_user_xattr(fs_context, file_path, XATTR_NOOBAA_CUSTOM_PREFIX);
         } catch (err) {
             dbg.error(`NamespaceFS.delete_object_tagging: failed in dir ${file_path} with error: `, err);
-            throw this._translate_object_error_codes(err);
+            throw native_fs_utils.translate_error_codes(err, native_fs_utils.entity_enum.OBJECT);
         }
         return {version_id: params.version_id};
     }
@@ -1992,7 +1992,7 @@ class NamespaceFS {
             await this.set_fs_xattr_op(fs_context, file_path, fs_xattr, undefined);
         } catch (err) {
             dbg.error(`NamespaceFS.put_object_tagging: failed in dir ${file_path} with error: `, err);
-            throw this._translate_object_error_codes(err);
+            throw native_fs_utils.translate_error_codes(err, native_fs_utils.entity_enum.OBJECT);
         }
         return { tagging: [], version_id: params.version_id };
     }
@@ -2121,7 +2121,7 @@ class NamespaceFS {
             }
         } catch (error) {
             dbg.error('namespace_fs.restore_object: failed with error: ', error, file_path);
-            throw this._translate_object_error_codes(error);
+            throw native_fs_utils.translate_error_codes(error, native_fs_utils.entity_enum.OBJECT);
         } finally {
             if (file) await file.close(fs_context);
         }
@@ -2200,7 +2200,7 @@ class NamespaceFS {
             file = null;
         } catch (error) {
             dbg.error('NamespaceFS.handle_fs_xattr_op: failed with error: ', error, file_path);
-            throw this._translate_object_error_codes(error);
+            throw native_fs_utils.translate_error_codes(error, native_fs_utils.entity_enum.OBJECT);
         } finally {
             if (file) await file.close(fs_context);
         }
@@ -2372,21 +2372,12 @@ class NamespaceFS {
         }
     }
 
-    _translate_object_error_codes(err) {
-        if (err.rpc_code) return err;
-        if (err.code === 'ENOENT' || err.code === 'ENOTDIR') err.rpc_code = 'NO_SUCH_OBJECT';
-        if (err.code === 'EEXIST') err.rpc_code = 'BUCKET_ALREADY_EXISTS';
-        if (err.code === 'EPERM' || err.code === 'EACCES') err.rpc_code = 'UNAUTHORIZED';
-        if (err.code === 'IO_STREAM_ITEM_TIMEOUT') err.rpc_code = 'IO_STREAM_ITEM_TIMEOUT';
-        if (err.code === 'INTERNAL_ERROR') err.rpc_code = 'INTERNAL_ERROR';
-        return err;
-    }
-
     async _load_bucket(params, fs_context) {
         try {
             await nb_native().fs.stat(fs_context, this.bucket_path);
         } catch (err) {
-            throw this._translate_object_error_codes(err);
+            dbg.warn('_load_bucket failed, on bucket_path', this.bucket_path, 'got error', err);
+            throw native_fs_utils.translate_error_codes(err, native_fs_utils.entity_enum.BUCKET);
         }
     }
 
@@ -2451,7 +2442,9 @@ class NamespaceFS {
         try {
             await nb_native().fs.mkdir(fs_context, params.full_path, native_fs_utils.get_umasked_mode(0o777));
         } catch (err) {
-            throw this._translate_object_error_codes(err);
+            dbg.error('NamespaceFS: create_uls fs_context:', fs_context, 'new_dir_path: ',
+                params.full_path, 'got an error:', err);
+            throw native_fs_utils.translate_error_codes(err, native_fs_utils.entity_enum.BUCKET);
         }
     }
 
@@ -2468,7 +2461,9 @@ class NamespaceFS {
 
             await native_fs_utils.folder_delete(params.full_path, fs_context);
         } catch (err) {
-            throw this._translate_object_error_codes(err);
+            dbg.error('NamespaceFS: delete_uls fs_context:', fs_context, 'to_delete_dir_path: ',
+                params.full_path, 'got an error:', err);
+            throw native_fs_utils.translate_error_codes(err, native_fs_utils.entity_enum.BUCKET);
         }
     }
 

--- a/src/test/unit_tests/test_bucketspace_fs.js
+++ b/src/test/unit_tests/test_bucketspace_fs.js
@@ -11,7 +11,8 @@ const assert = require('assert');
 const P = require('../../util/promise');
 const config = require('../../../config');
 const fs_utils = require('../../util/fs_utils');
-const { get_process_fs_context, read_file, get_user_by_distinguished_name, get_bucket_tmpdir_name } = require('../../util/native_fs_utils');
+const { get_process_fs_context, read_file, get_user_by_distinguished_name, get_bucket_tmpdir_name,
+    update_config_file } = require('../../util/native_fs_utils');
 const nb_native = require('../../util/nb_native');
 const SensitiveString = require('../../util/sensitive_string');
 const NamespaceFS = require('../../sdk/namespace_fs');
@@ -22,6 +23,7 @@ const nc_mkm = require('../../manage_nsfs/nc_master_key_manager').get_instance()
 
 
 const test_bucket = 'bucket1';
+const test_bucket2 = 'bucket2';
 const test_not_empty_bucket = 'notemptybucket';
 const test_bucket_temp_dir = 'buckettempdir';
 const test_bucket_invalid = 'bucket_invalid';
@@ -507,6 +509,43 @@ mocha.describe('bucketspace_fs', function() {
             // account_iam_user2 can delete the created bucket (the implicit policy - same root account)
             const dummy_object_sdk_for_account_iam_user2 = make_dummy_object_sdk_for_account(dummy_object_sdk, account_iam_user2);
             await bucketspace_fs.delete_bucket(param, dummy_object_sdk_for_account_iam_user2);
+        });
+
+        mocha.it('delete_bucket after the ULS was deleted (should_create_underlying_storage true)', async function() {
+            const param = { name: test_bucket2 };
+            await create_bucket(param.name);
+
+            // this is not a case that we want, but it might happen: if somehow the ULS (Underline Storage) was deleted
+            const uls_path = path.join(new_buckets_path, param.name);
+            await fs.promises.stat(uls_path);
+            await fs.promises.rm(uls_path, { recursive: true }); // somehow the ULS is deleted
+
+            await bucketspace_fs.delete_bucket(param, dummy_object_sdk);
+            const bucket_config_path = get_config_file_path(CONFIG_SUBDIRS.BUCKETS, param.name);
+            await fs_utils.file_must_not_exist(bucket_config_path);
+        });
+
+        mocha.it('delete_bucket after the ULS was deleted (should_create_underlying_storage false)', async function() {
+            const param = { name: test_bucket2 };
+            await create_bucket(param.name);
+
+            // we want to mock a bucket creation using the CLI,
+            // we manually change the should_create_underlying_storage to false
+            const bucket_config_path = get_config_file_path(CONFIG_SUBDIRS.BUCKETS, param.name);
+            const data = await fs.promises.readFile(bucket_config_path);
+            const bucket = await JSON.parse(data.toString());
+            assert.ok(bucket.should_create_underlying_storage === true);
+            bucket.should_create_underlying_storage = false;
+            await update_config_file(process_fs_context, CONFIG_SUBDIRS.BUCKETS, bucket_config_path, JSON.stringify(bucket));
+            await fs_utils.file_must_exist(bucket_config_path);
+
+            // this is not a case that we want, but it might happen: if somehow the ULS (Underline Storage) was deleted
+            const uls_path = path.join(new_buckets_path, param.name);
+            await fs.promises.stat(uls_path);
+            await fs.promises.rm(uls_path, { recursive: true }); // somehow the ULS is deleted
+
+            await bucketspace_fs.delete_bucket(param, dummy_object_sdk);
+            await fs_utils.file_must_not_exist(bucket_config_path);
         });
     });
     mocha.describe('set_bucket_versioning', function() {

--- a/src/util/native_fs_utils.js
+++ b/src/util/native_fs_utils.js
@@ -19,6 +19,14 @@ const VALID_BUCKET_NAME_REGEXP = /^(([a-z0-9]|[a-z0-9][a-z0-9-]*[a-z0-9])\.)*([a
 
 /** @typedef {import('../util/buffer_utils').MultiSizeBuffersPool} MultiSizeBuffersPool */
 
+// we will use this object to determined the entity for the rpc_code error message 
+const entity_enum = Object.freeze({
+    OBJECT: 'OBJECT',
+    BUCKET: 'BUCKET',
+    USER: 'USER',
+    ACCESS_KEY: 'ACCESS_KEY',
+});
+
 function get_umasked_mode(mode) {
     // eslint-disable-next-line no-bitwise
     return mode & ~config.NSFS_UMASK;
@@ -580,6 +588,25 @@ function get_bucket_tmpdir_full_path(bucket_path, bucket_id) {
     return path.join(bucket_path, get_bucket_tmpdir_name(bucket_id));
 }
 
+
+/**
+ * translate_error_codes we translate FS error codes to rpc_codes (strings)
+ * and add the rpc_code property to the original error object
+ * @param {object} err
+ * @param {('OBJECT'|'BUCKET'|'USER'|'ACCESS_KEY')} entity
+ */
+function translate_error_codes(err, entity) {
+    if (err.rpc_code) return err;
+    if (err.code === 'ENOENT' || err.code === 'ENOTDIR') {
+        err.rpc_code = `NO_SUCH_${entity}`;
+    }
+    if (err.code === 'EEXIST') err.rpc_code = `${entity}_ALREADY_EXISTS`;
+    if (err.code === 'EPERM' || err.code === 'EACCES') err.rpc_code = 'UNAUTHORIZED';
+    if (err.code === 'IO_STREAM_ITEM_TIMEOUT') err.rpc_code = 'IO_STREAM_ITEM_TIMEOUT';
+    if (err.code === 'INTERNAL_ERROR') err.rpc_code = 'INTERNAL_ERROR';
+    return err;
+}
+
 exports.get_umasked_mode = get_umasked_mode;
 exports._make_path_dirs = _make_path_dirs;
 exports._create_path = _create_path;
@@ -615,3 +642,5 @@ exports.is_dir_rw_accessible = is_dir_rw_accessible;
 exports.folder_delete = folder_delete;
 exports.get_bucket_tmpdir_full_path = get_bucket_tmpdir_full_path;
 exports.get_bucket_tmpdir_name = get_bucket_tmpdir_name;
+exports.entity_enum = entity_enum;
+exports.translate_error_codes = translate_error_codes;

